drivers/include/net/netdev.h

```diff
@@ -300,75 +307,118 @@ struct netdev {
-     * @brief Send frame
+     * @brief   Start transmission of the given frame and return directly
      *
-     * @pre `(dev != NULL) && (iolist != NULL`
+     * @pre     `(dev != NULL) && (iolist != NULL)`
      *
-     * @param[in] dev       Network device descriptor. Must not be NULL.
-     * @param[in] iolist    IO vector list to send. Elements of this list may
-     *                      have iolist_t::iol_data == NULL or
-     *                      iolist_t::iol_size == 0. However, unless otherwise
-     *                      specified by the device, the *first* element
-     *                      must contain data.
+     * @param[in]   dev     Network device descriptor. Must not be NULL.
+     * @param[in]   iolist  IO vector list to send. Elements of this list may
+     *                      have iolist_t::iol_size == 0 and (in this case only)
+     *                      iolist_t::iol_data == 0.
      *
-     * @return negative errno on error
-     * @return number of bytes sent
+     * @retval  -EBUSY      Driver is temporarily unable to send, e.g. because
+     *                      an incoming frame on a half-duplex medium is
+     *                      received
+     * @retval  -ENETDOWN   Device is powered down
+     * @retval  <0          Other error
+     * @retval  0           Transmission successfully started
+     * @retval  >0          Number of bytes transmitted (deprecated!)
+     *
+     * This function will cause the driver to start the transmission in an
+     * async fashion. The driver will "own" the `iolist` until a subsequent
+     * call to @ref netdev_driver_t::confirm_send returns `0`. The driver must
+     * signal completion using the NETDEV_EVENT_TX_COMPLETE event, regardless
+     * of success or failure.
+     *
+     * Old drivers might not be ported to the new API and have
+     * netdev_driver_t::confirm_send set to `NULL`. An that case the driver
+     * will return the number of bytes transmitted on success (instead of `0`)
+     * and will likely block until completion.
      */
     int (*send)(netdev_t *dev, const iolist_t *iolist);
 
     /**
-     * @brief Drop a received frame, **OR** get the length of a received frame,
-     *        **OR** get a received frame.
+     * @brief   Fetch the status of a transmission and perform any potential
+     *          cleanup
+     *
+     * @param[in]   dev     Network device descriptor. Must not be NULL.
+     * @param[out]  info    Device class specific type to fetch transmission
+     *                      info. May be `NULL` if not needed by upper layer.
+     *                      May be ignored by driver.
+     *
+     * @return  Number of bytes transmitted. (The size of the transmitted
+     *          frame including all overhead, such as frame check sequence,
+     *          bit stuffing, escaping, headers, trailers, preambles, start of
+     *          frame delimiters, etc. May be an estimate for performance
+     *          reasons.)
```

Comment:
@miri64: We had a discussion on what this size should be in an issue. This interpretation is different to what most drivers implement (in the `send()`, not the `confirm_send()` function), but as porting would result in work anyway, I figured we should rather go for the best solution than the lazy solution.

IMO there is little reason to just return the result of `iolist_size(iolist)` on what was passed in `send()`, as the upper layer could just call `iolist_size()` directly to get this. So if we really want additional data for statistics, it should be IMO a value that has the most value when the value of `iolist_size(iolist)` is known anyway. And to me this is the number of bytes physically transmitted; as this would allow to calculate the layer 2 throughput, while `iolist_size(iolist)` can provide the layer 2 goodput.

Reply:
The last sentence is wrong. The netif can easily calculate the sizes needed for goodput calculation, but it would need to make sure to call `iolist_size(iolist)` prior to adding the L2 header (and trailer, if applicable). But the reasoning should still be valid.